core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,570 @@
+ *   2. Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in
+ *      the documentation and/or other materials provided with the
+ *      distribution.
+ *   3. Neither the name of Intel Corporation nor the names of its
+ *      contributors may be used to endorse or promote products derived
+ *      from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "enclave_hsm_t.h"
+#include "log_utils.h"
+#include "sgx_tseal.h"
+
+#include <string>
+#include <stdio.h>
+#include <stdbool.h>
+#include <mbusafecrt.h>
+
+#include "sgx_report.h"
+#include "sgx_utils.h"
+#include "sgx_tkey_exchange.h"
+
+#include "datatypes.h"
+#include "openssl/rsa.h"
+#include "openssl/evp.h"
+#include "openssl/ec.h"
+#include "openssl/pem.h"
+#include "openssl/bio.h"
+#include "openssl/rand.h"
+#include "openssl/err.h"
+
+#include "key_factory.h"
+#include "key_operation.h"
+
+#define SGX_DOMAIN_KEY_SIZE 16
+
+#define RSA_2048_KEY_BITS 2048
+#define RSA_3072_KEY_BITS 3072
+#define RSA_4096_KEY_BITS 4096
+
+#define ECC_MAX_PLAINTEXT_SIZE 256
+
+sgx_aes_gcm_128bit_key_t g_domain_key = {0};
+
+using namespace std;
+
+uint32_t ehsm_calc_keyblob_len(const uint32_t keyspec)
+{
+    switch (keyspec)
+    {
+    case EH_RSA_2048:
+        return PEM_BUFSIZE * 3 + sizeof(sgx_aes_gcm_data_ex_t);
+        break;
+    case EH_RSA_3072:
+        return PEM_BUFSIZE * 4 + sizeof(sgx_aes_gcm_data_ex_t);
+        break;
+    case EH_RSA_4096:
+        return PEM_BUFSIZE * 5 + sizeof(sgx_aes_gcm_data_ex_t);
+        break;        
+    case EH_EC_P224:
+    case EH_SM2:
+        return PEM_BUFSIZE + sizeof(sgx_aes_gcm_data_ex_t);
+        break;
+    case EH_AES_GCM_128:
+    case EH_SM4_CTR:
+    case EH_SM4_CBC:
+        return 16 + sizeof(sgx_aes_gcm_data_ex_t);
+    case EH_AES_GCM_192:
+        return 24 + sizeof(sgx_aes_gcm_data_ex_t);
+    case EH_AES_GCM_256:
+        return 32 + sizeof(sgx_aes_gcm_data_ex_t);
+    default:
+        return UINT32_MAX;
+        break;
+    }
+}
+
+uint32_t ehsm_get_gcm_ciphertext_size(const sgx_aes_gcm_data_ex_t *gcm_data)
+{
```

Comment:
what's this for? seems useless

Reply:
This encapsulates a structure pointer conversion, which can save some code